fem/quadinterpolator_det.cpp

```diff
@@ -0,0 +1,221 @@
+// CONTRIBUTING.md for details.
+
+#include "quadinterpolator.hpp"
+#include "../general/forall.hpp"
+#include "../linalg/dtensor.hpp"
+#include "../fem/kernels.hpp"
+#include "../linalg/kernels.hpp"
+
+using namespace mfem;
+
+namespace mfem
+{
+
+template<int T_D1D = 0, int T_Q1D = 0, int MAX_D1D = 0, int MAX_Q1D = 0>
+static void Det2D(const int NE,
+                  const double *b,
+                  const double *g,
+                  const double *x,
+                  double *y,
+                  const int vdim = 1,
+                  const int d1d = 0,
+                  const int q1d = 0)
+{
+   constexpr int DIM = 2;
+   static constexpr int NBZ = 1;
+
+   const int D1D = T_D1D ? T_D1D : d1d;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+
+   const auto B = Reshape(b, Q1D, D1D);
+   const auto G = Reshape(g, Q1D, D1D);
+   const auto X = Reshape(x,  D1D, D1D, DIM, NE);
+   auto Y = Reshape(y, Q1D, Q1D, NE);
+
+   MFEM_FORALL_2D(e, NE, Q1D, Q1D, NBZ,
+   {
+      constexpr int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+      constexpr int MD1 = T_D1D ? T_D1D : MAX_D1D;
+      const int D1D = T_D1D ? T_D1D : d1d;
+      const int Q1D = T_Q1D ? T_Q1D : q1d;
+
+      MFEM_SHARED double BG[2][MQ1*MD1];
+      MFEM_SHARED double XY[2][NBZ][MD1*MD1];
+      MFEM_SHARED double DQ[4][NBZ][MD1*MQ1];
+      MFEM_SHARED double QQ[4][NBZ][MQ1*MQ1];
+
+      kernels::LoadX<MD1,NBZ>(e,D1D,X,XY);
+      kernels::LoadBG<MD1,MQ1>(D1D,Q1D,B,G,BG);
+
+      kernels::GradX<MD1,MQ1,NBZ>(D1D,Q1D,BG,XY,DQ);
+      kernels::GradY<MD1,MQ1,NBZ>(D1D,Q1D,BG,DQ,QQ);
+
+      MFEM_FOREACH_THREAD(qy,y,Q1D)
+      {
+         MFEM_FOREACH_THREAD(qx,x,Q1D)
+         {
+            double J[4];
+            kernels::PullGrad<MQ1,NBZ>(qx,qy,QQ,J);
+            Y(qx,qy,e) = kernels::Det<2>(J);
+         }
+      }
+   });
+}
+
+template<int T_D1D = 0, int T_Q1D = 0, int MAX_D1D = 0, int MAX_Q1D = 0,
+         bool SMEM = true>
+static void Det3D(const int NE,
+                  const double *b,
+                  const double *g,
+                  const double *x,
+                  double *y,
+                  const int vdim = 1,
+                  const int d1d = 0,
+                  const int q1d = 0)
+{
+   constexpr int DIM = 3;
+   static constexpr int MQ1 = T_Q1D ? T_Q1D : MAX_Q1D;
+   static constexpr int MD1 = T_D1D ? T_D1D : MAX_D1D;
+   static constexpr int MDQ = MQ1 > MD1 ? MQ1 : MD1;
+   static constexpr int MSZ = MDQ * MDQ * MDQ * 9;
+   static constexpr int GRID = SMEM ? 0 : 128;
+
+   const int D1D = T_D1D ? T_D1D : d1d;
+   const int Q1D = T_Q1D ? T_Q1D : q1d;
+
+   const auto B = Reshape(b, Q1D, D1D);
+   const auto G = Reshape(g, Q1D, D1D);
+   const auto X = Reshape(x, D1D, D1D, D1D, DIM, NE);
+   auto Y = Reshape(y, Q1D, Q1D, Q1D, NE);
+
+   double *GM = nullptr;
+   if (!SMEM)
+   {
+      static Vector gmem;
+      gmem.SetSize(2*MSZ*GRID);
+      gmem.UseDevice(true);
+      GM = gmem.Write();
```

Comment:
We should probably avoid global device allocations like this (especially in template functions where each instantiation will get its own copy of `gmem`). It will be better to make this allocation as part of the object(s) that use this kernel -- that way the memory will be freed when the object is destroyed.

Reply:
I pushed a small update (951d55c) to address this.

Reply:
I just added the `d_buffer` to `UseDevice(true)`

Reply:
I didn't keep that because `d_buff->Write()` does that too.